### PR TITLE
Text Gen slice decoder projection optimisation

### PR DESCRIPTION
**Pull request by jimypbr**
# What does this PR do?

Optimization of the decoder-only forward mode used in text generation.
Only computes the projection on the position `t` of the sequence rather than on the whole sequence.
This saves a lot of memory in the LM-head matmul.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/graphcore/generation_utils.py [hunk 1]
```diff
@@ -81,12 +102,15 @@ def forward(self, t, **model_inputs):
         Returns:
             The output logits at position `t` only
         """
+        # Update the cur_step buffer in the _SliceLinear layer
+        self.pipelined_model.get_output_embeddings().cur_step.copy_(t)
```
Comment (katalinic-gc) [inline optimum/graphcore/generation_utils.py:106]: I am still contemplating if there's a cleaner way to do this

File: optimum/graphcore/generation_utils.py [hunk 2]
```diff
@@ -63,6 +63,25 @@ def graph_profile_dir_append(append: str):
...
+    `self.cur_step` is a buffer.
+    """
+
+    def __init__(self, linear_layer):
+        super().__init__()
+        self.wrapped_linear = linear_layer
+        self.register_buffer("cur_step", torch.tensor([0], dtype=torch.int32))
+
+    def forward(self, x):
+        x = poptorch.dynamic_slice(x, 1, self.cur_step, 1, 1)
```
Comment (katalinic-gc) [inline optimum/graphcore/generation_utils.py:81]: Just wondering, will seq length always be in axis 1? Do we ensure in beam search e.g. that beams go into batch dim?

File: optimum/graphcore/generation_utils.py [hunk 3]
```diff
@@ -63,6 +63,25 @@ def graph_profile_dir_append(append: str):
...
+    before executing the linear. The intended use is as an optimized replacement of the
+    LM Head in the Decoder for text generation inference.
+    The slice is performed on the position `self.cur_step` of the input tensor.
+    `self.cur_step` is a buffer.
+    """
+
+    def __init__(self, linear_layer):
+        super().__init__()
+        self.wrapped_linear = linear_layer
+        self.register_buffer("cur_step", torch.tensor([0], dtype=torch.int32))
```
[the remark this answers is not in the document]
Comment (jimypbr) [reply optimum/graphcore/generation_utils.py]: yeah I've tested this with gpt2, bart, and t5 and they work as before.
